ath79: add support for MikroTik hAP ac

rmounce opened the pull request:
This patch adds support for the MikroTik hAP ac (RouterBOARD 962UiGS-5HacT2HnT)

Working:
- 2.4GHz WLAN (SoC)
- 5GHz WLAN (QCA9880)
- Switch (QCA8337)
- SFP cage (tested with genuine & generic GLC-T)
- USB

Not working:
- Red WLAN LED

Note that config persistence on sysupgrade only worked after applying https://github.com/openwrt/openwrt/pull/3271

rmounce committed:
ath79: add support for MikroTik RouterBOARD 962UiGS-5HacT2HnT (hAP ac)

This patch adds support for the MikroTik RouterBOARD 962UiGS-5HacT2HnT (hAP ac)

Specifications:
- SoC: QCA9558
- RAM: 128 MB
- Flash: 16 MB SPI
- 2.4GHz WLAN: 3x3:3 802.11n on SoC
- 5GHz WLAN: 3x3:3 802.11ac on QCA9880 connected via PCIe
- Switch: 5x 1000/100/10 on QCA8337 connected via RGMII
- SFP cage: connected via SGMII (tested with genuine & generic GLC-T)
- USB: 1x type A, GPIO power switch
- PoE: Passive input on Ether1, GPIO switched passthrough to Ether5
- Reset button
- "SFP" LED connected to SoC
- Ethernet LEDs connected to QCA8337 switch
- Green WLAN LED connected to QCA9880

Not working:
- Red WLAN LED

Installation:
TFTP boot initramfs image and then perform sysupgrade. Follow common
MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Ryan Mounce <ryan@mounce.com.au>

zero13th commented:
Please remove this if it's not allowed to add reference: adron-s/openwrt-rb962@7520904

adschm commented:
This has formal issues. Please add specifications and flashing instructions to the commit message.

john-tho commented:
Thank you for submitting a PR for this device.

> Note that config persistence on sysupgrade only worked after applying #3271

Sysupgrade does not work currently (in master, or 21.02) on any of the ath79 mikrotik NOR devices. https://bugs.openwrt.org/index.php?do=details&task_id=3492. I would like my PR reviewed, but removing the `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS` ath79 mikrotik kernel config should be much simpler? Or maybe setting a blocksize in the image recipe is an alternative?
```
--- a/target/linux/ath79/mikrotik/config-default
+++ b/target/linux/ath79/mikrotik/config-default
@@ -18,7 +18,6 @@ CONFIG_MTD_NAND_RB4XX=y
 CONFIG_MTD_RAW_NAND=y
 CONFIG_MTD_ROUTERBOOT_PARTS=y
 CONFIG_MTD_SPI_NAND=y
-CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y
 CONFIG_MTD_SPLIT_MINOR_FW=y
 CONFIG_MTD_UBI=y
 CONFIG_MTD_UBI_BLOCK=y
```

Cheers

adschm commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-962uigs-5hact2hnt.dts:
```diff
@@ -0,0 +1,227 @@
+
+/ {
+	compatible = "mikrotik,routerboard-962uigs-5hact2hnt", "qca,qca9558";
+	model = "MikroTik hAP ac";
```
model and compatible should be the same.

adschm commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-962uigs-5hact2hnt.dts:
```diff
@@ -0,0 +1,227 @@
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_usb_power {
```
just usb_power should be enough.

adschm commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-962uigs-5hact2hnt.dts:
```diff
@@ -0,0 +1,227 @@
+		compatible = "gpio-export";
+
+		gpio_usb_power {
+			gpio-export,name = "power:usb";
```
I'd not use the colon-syntax here, but just something like power-usb or usb-power.

rmounce replied:
This block was copied from `qca9558_mikrotik_routerboard-922uags-5hpacd.dts`. I agree with your syntax, however would it be preferable to have consistent labels across similar boards?

adschm replied:
Why do you resolve this block if you ask a question? That makes no sense at all and just looks like you want to hide open problems from reviewers.

Coming back to the subject: This train has been missed a long time ago. gpio-export names are inherently inconsistent. But if you prefer the ":" syntax, I will not stop you from using it. But be aware that one cannot change this name without migration, so choose wisely.

rmounce replied:
Changed per your feedback, marked resolved. Then reconsidered, sought clarification and forgot to un-resolve. If I wanted to hide problems I would not ask any questions :)

I don't have strong opinions on syntax here, seems we are both happy with `usb-power`.

adschm commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-962uigs-5hact2hnt.dts:
```diff
@@ -0,0 +1,227 @@
+		};
+
+		poe_passthrough_port5 {
+			gpio-export,name = "mikrotik:poe-passthrough:port5";
```
No need to add mikrotik reference, consider removing the colon here as well.

adschm commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-962uigs-5hact2hnt.dts:
```diff
@@ -0,0 +1,227 @@
+		tx-disable-gpios = <&gpio 16 GPIO_ACTIVE_HIGH>;
+		// Toggling GPIO16 actually enables/disables the transmitter,
+		// but the SFP driver does not seem to be using it.
+		};
```
wrong indent.

adschm commented on target/linux/ath79/image/mikrotik.mk:
```diff
@@ -28,6 +28,17 @@ define Device/mikrotik_routerboard-922uags-5hpacd
+define Device/mikrotik_routerboard-962uigs-5hact2hnt
+  $(Device/mikrotik_nor)
+  SOC := qca9558
+  DEVICE_MODEL := RouterBOARD 962UiGS-5HacT2HnT (hAP ac)
```
That's what should be used for the DTS model as well (+ Ubiquiti ...) ...

adschm commented on target/linux/ath79/image/mikrotik.mk:
```diff
@@ -28,6 +28,17 @@ define Device/mikrotik_routerboard-922uags-5hpacd
+  SOC := qca9558
+  DEVICE_MODEL := RouterBOARD 962UiGS-5HacT2HnT (hAP ac)
+  DEVICE_PACKAGES += kmod-ath10k-ct ath10k-firmware-qca988x-ct kmod-usb2 \
+  kmod-i2c-gpio kmod-sfp
```
hanging indent one tab ...

adschm commented on target/linux/ath79/image/mikrotik.mk:
```diff
@@ -28,6 +28,17 @@ define Device/mikrotik_routerboard-922uags-5hpacd
+  DEVICE_PACKAGES += kmod-ath10k-ct ath10k-firmware-qca988x-ct kmod-usb2 \
+  kmod-i2c-gpio kmod-sfp
+  IMAGE_SIZE := 16256k
+  SUPPORTED_DEVICES += rb-962uigs-5hact2hnt
```
Have you checked that this is the correct string for ar71xx?

rmounce replied:
Yes.

adschm commented on target/linux/ath79/mikrotik/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -31,6 +31,9 @@ case "$FIRMWARE" in
 	mikrotik,routerboard-wap-g-5hact2hnd)
 		caldata_mikrotik_ath9k 0x1000 0x440 $(macaddr_add "$mac_base" +2)
 		;;
+	mikrotik,routerboard-962uigs-5hact2hnt)
```
Sorting ...

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,10 @@ ath79_setup_interfaces()
 		ucidef_add_switch "switch1" \
 			"0@eth1" "1:lan:4" "2:lan:1" "3:lan:2" "4:lan:3"
 		;;
+	mikrotik,routerboard-962uigs-5hact2hnt)
```
Sorting.

pmelange commented:
@adron-s has also started adding support for the rb960pgs.  In doing so, he moved a large part of the dts into a qca9558_mikrotik_routerboard-96x.dtsi file which is common to both rb962uigs and the rb960pgs.  (https://github.com/adron-s/openwrt-rb962/commit/458d0f01efb4a6dd6ff86d4f0bd442fadd247b3b)

I think it may be a good idea to have the dtsi file in this patchset as well.

rmounce commented:
> @adron-s has also started adding support for the rb960pgs. In doing so, he moved a large part of the dts into a qca9558_mikrotik_routerboard-96x.dtsi file which is common to both rb962uigs and the rb960pgs. ([adron-s/openwrt-rb962@24b26f3](https://github.com/adron-s/openwrt-rb962/commit/24b26f39766f7d0712bba5cc1a00bf377dfa931e))
> 
> I think it may be a good idea to have the dtsi file in this patchset as well.

I've split this into dtsi/dts and made some other changes now that I've found that both @adron-s and @john-tho both have WIP rb926 branches. I am interested to know how they got dumps of OEM register values & comprehensive GPIO assignments. I have an old GPL source drop from Mikrotik that is not so comprehensive.

Some general notes:
- Full ath10k GPIO driver support does not appear to be implemented anywhere to be able to control red WLAN LED & power LEDs, despite some devicetree configuration suggesting this capability. `201-ath10k-add-LED-and-GPIO-controlling-support-for-various-chipsets.patch` maps the 'default' green WLAN LED for QCA9880 as `ath10k-phy0` without any devicetree configuration, so I have removed some non-functional snippets that I had copied.
- In my testing, both `pcie0` and `pcie1` nodes must be enabled for QCA9880 firmware load to succeed?!

john-tho commented on target/linux/ath79/mikrotik/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -19,7 +19,8 @@ case "$FIRMWARE" in
 	;;
 "ath10k/cal-pci-0000:01:00.0.bin")
 	case $board in
-	mikrotik,routerboard-922uags-5hpacd)
+	mikrotik,routerboard-922uags-5hpacd|\
+	mikrotik,routerboard-962uigs-5hact2hnt)
 		caldata_sysfsload_from_file $wlan_data 0x5000 0x844
```
Regarding needing both pcie0 & pcie1 enabled: I _think_ when DTS SoC pcie0 is disabled, SoC pcie1 becomes linux pcie0, so this 962 board case needs to be in a different FIRMWARE case?

rmounce replied:
You're exactly right. PR updated.

john-tho commented:
For information, my 962 GPIO details came from:
- Some from the old Mikrotik GPL dump
- Others from:
  - [Find SoC registers from datasheet or drivers](https://github.com/Deoptim/atheros)
  - Jailbreak OEM firmware
  - Watch (`busybox devmem`) the memory locations of the GPIO hardware registers while individually triggering each setting or hardware change that is expected to change a GPIO.

tomjnixon commented:
I just started testing this, and I found that the initramfs image doesn't boot on my board. Using the lzma loader seems to sort it, for some reason:

```patch
diff --git a/target/linux/ath79/image/mikrotik.mk b/target/linux/ath79/image/mikrotik.mk
index 3df905d161..6353cc6840 100644
--- a/target/linux/ath79/image/mikrotik.mk
+++ b/target/linux/ath79/image/mikrotik.mk
@@ -45,6 +45,8 @@ define Device/mikrotik_routerboard-962uigs-5hact2hnt
        kmod-i2c-gpio kmod-sfp
   IMAGE_SIZE := 16256k
   SUPPORTED_DEVICES += rb-962uigs-5hact2hnt
+  LOADER_TYPE := elf
+  KERNEL := kernel-bin | append-dtb-elf | lzma | loader-kernel
 endef
 TARGET_DEVICES += mikrotik_routerboard-962uigs-5hact2hnt
```
I'm not sure why this happens, and I don't get any serial output at all, but this used to be used on all mikrotik images. Perhaps this is down to using an older bootloader version? I'm on 3.34 according to `/sys/firmware/mikrotik/hard_config/booter_version`.

john-tho commented:
On Sun, 18 Jul 2021, at 18:08, Thomas Nixon wrote:
> 
> 
> I just started testing this, and I found that the initramfs image 
> doesn't boot on my board. Using the lzma loader seems to sort it, for 
> some reason:
> 
> diff --git a/target/linux/ath79/image/mikrotik.mk 
> b/target/linux/ath79/image/mikrotik.mk
> index 3df905d161..6353cc6840 100644
> --- a/target/linux/ath79/image/mikrotik.mk
> +++ b/target/linux/ath79/image/mikrotik.mk
> @@ -45,6 +45,8 @@ define Device/mikrotik_routerboard-962uigs-5hact2hnt
>         kmod-i2c-gpio kmod-sfp
>    IMAGE_SIZE := 16256k
>    SUPPORTED_DEVICES += rb-962uigs-5hact2hnt
> +  LOADER_TYPE := elf
> +  KERNEL := kernel-bin | append-dtb-elf | lzma | loader-kernel
>  endef
>  TARGET_DEVICES += mikrotik_routerboard-962uigs-5hact2hnt
> I'm not sure why this happens, and I don't get any serial output at 
> all, but this used to be used on all mikrotik images. Perhaps this is 
> down to using an older bootloader version? I'm on 3.34 according to 
> `/sys/firmware/mikrotik/hard_config/booter_version`.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openwrt/openwrt/pull/4176#issuecomment-882096514>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AD4G7K5OEISYA5S7I3TGH4TTYMKDJANCNFSM4453YNEA>.
> 

Hi Thomas,

If you are getting no console output, it is likely that RouterBoot is not starting the initramfs.
There is a tag in RouterBoot hardconfg that you can change to make RouterBoot use console,
but it does not provide much information.

RouterBoot has a limit to both the netboot payload size, and the payload ELF load address.
The loader can relocate the kernel, whereas the self-extracting MIPS kernel cannot,
so a larger kernel initram image causes a bigger load address as well as a bigger filesize
for the self-extracting kernel.
If RouterBoot is failing to netboot a minimal initramfs, or the buildbot config initramfs image,
this is something that needs to be addressed. It if is a big image, I tend to build a minimal config initramfs,
then build my full image for the squashfs.

I moved away from the loader because it appeared to be causing issues on mt7621.

I am happy to continue discussing or diagnosing.

Cheers,
-- 
  John Thomson

rogerpueyo commented:
Hi,

> Note that config persistence on sysupgrade only worked after applying #3271

As a workaround, and since the SPI NOR devices are using 4k sectors, generating the images with 4k blocksize would be a workaround:
```
define Device/mikrotik_nor
  $(Device/mikrotik)
  BLOCKSIZE = 4k
  IMAGE/sysupgrade.bin := append-kernel | kernel2minor -s 1024 -e | \
	pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
	check-size | append-metadata
endef
```
It does not solve the 4k/64k issue, that's true, and the write speed is still slow, but settings will persist a sysupgrade.

Any comments? I didn't see this was suggested before, sorry if I'm bringing back something already discussed.

tomjnixon commented:
Hi,

> If you are getting no console output, it is likely that RouterBoot is not starting the initramfs. There is a tag in RouterBoot hardconfg that you can change to make RouterBoot use console, but it does not provide much information.

Sorry about this, I realised after posting that I'd messed up my build / testing and it was actually working fine; I deleted my post but forgot that some people use email. Thanks for the info though, I'll continue testing later.

john-tho commented:
> As a workaround, and since the SPI NOR devices are using 4k sectors, generating the images with 4k blocksize would be a workaround:
> 
> ```
> define Device/mikrotik_nor
>   $(Device/mikrotik)
>   BLOCKSIZE = 4k
>   IMAGE/sysupgrade.bin := append-kernel | kernel2minor -s 1024 -e | \
> 	pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
> 	check-size | append-metadata
> endef
> ```

Hi @rogerpueyo,

Looks fine. I would suggest you submit this a as separate patch so that it gets done sooner.

adron-s commented:
16.10.2021 15:21, pmelange пишет:
>
> Is there any progress in getting this device supported? Once done, I 
> would like to try to get the rb960pgs support as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openwrt/openwrt/pull/4176#issuecomment-944907399>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEEYPFYWNNOKWQFPIOYDMATUHFU4RANCNFSM4453YNEA>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>
>
Hello. I am using these two devices with the latest version of OpenWRT 
(21.02) and DSA. And they work great :-)

l29ah commented:
Sup?

danpawlik commented:
> @john-tho @adschm What needs to be done to this PR so that support for the rb962 gets merged? Is there anything that I can do help along?
> 
> Once merged, I would try to get the rb960pgs support done. I have this device and can test it.

I guess author needs to rebase that change to be up to date with current branch.

f00b4r0 commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -21,6 +21,10 @@ ath79_setup_interfaces()
 	mikrotik,routerboard-wapr-2nd)
 		ucidef_set_interface_lan "eth0"
 		;;
+	mikrotik,routerboard-962uigs-5hact2hnt)
+		ucidef_add_switch "switch0" \
+			"0@eth0" "2:lan" "3:lan" "4:lan" "5:lan" "1:wan"
+		;;
```
This does not define lan/wan: are the MAC addresses correct once booted?
How about port numbering  in Luci: is it correct as well?
Also what is eth1 used for since it does not appear here?

danpawlik replied:
Mac address is ok for LAN port and WAN (eth0.2). Is it SFP port eth0.1 or eth1?
```
  1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
      link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
      inet 127.0.0.1/8 scope host lo
         valid_lft forever preferred_lft forever
      inet6 ::1/128 scope host
         valid_lft forever preferred_lft forever
  2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
      link/ether aa:bd:2b:e9:01:25 brd ff:ff:ff:ff:ff:ff
      inet6 fe80::a8bd:2bff:fee9:125/64 scope link
         valid_lft forever preferred_lft forever
  3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
      link/ether 42:ee:c6:0b:3a:28 brd ff:ff:ff:ff:ff:ff
  6: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
      link/ether some:mac:address:b2 brd ff:ff:ff:ff:ff:ff
  7: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
      link/ether some:mac:address:ac brd ff:ff:ff:ff:ff:ff
      inet 192.168.88.1/24 brd 192.168.88.255 scope global br-lan
         valid_lft forever preferred_lft forever
      inet6 fded:7949:5479::1/60 scope global noprefixroute
         valid_lft forever preferred_lft forever
      inet6 fe80::764d:28ff:fecf:8cac/64 scope link
         valid_lft forever preferred_lft forever
  8: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
      link/ether some:mac:address:ac brd ff:ff:ff:ff:ff:ff
  9: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
      link/ether some:mac:address:ab brd ff:ff:ff:ff:ff:ff
      inet SOME IP ADDRESS.202/26 brd SOME IP ADDRESS.255 scope global eth0.2
         valid_lft forever preferred_lft forever
      inet6 fe80::764d:28ff:fecf:8cab/64 scope link
         valid_lft forever preferred_lft forever
  10: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
      link/ether some:mac:address:b1 brd ff:ff:ff:ff:ff:ff
      inet6 fe80::764d:28ff:fecf:8cb1/64 scope link
         valid_lft forever preferred_lft forever
```

eosita replied:
Hi Daniel. Following the output where you can see that both eth0 and eth1 have MAC different from the "real one" (the one written on the sticker, that starts with b8:69:f4:35:d6:74 and ends with b8:f4:35:d6:7b for WAN1).
I don't know exactly what this means and, also, what I have to do with the four lines of code you've written for "target/linux/ath79/mikrotik/base-files/etc/board.d/02_network"
If you think it's useful, I could reload the RouterOS on the hAP ac to write down the exact mac for the SFP module, even if I suppose it's b8:69:f4:35:d6:79.
Let me know if I could be useful for some testing. Thanks

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether 02:30:fc:e2:ea:26 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::30:fcff:fee2:ea26/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether f6:dd:03:43:fe:69 brd ff:ff:ff:ff:ff:ff
5: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:35:d6:7a brd ff:ff:ff:ff:ff:ff
6: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:35:d6:7b brd ff:ff:ff:ff:ff:ff
7: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:35:d6:75 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fd45:93f:f7fb::1/60 scope global noprefixroute 
       valid_lft forever preferred_lft forever
    inet6 fe80::ba69:f4ff:fe35:d675/64 scope link 
       valid_lft forever preferred_lft forever
8: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether b8:69:f4:35:d6:75 brd ff:ff:ff:ff:ff:ff
9: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:35:d6:74 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::ba69:f4ff:fe35:d674/64 scope link 
       valid_lft forever preferred_lft forever

danpawlik replied:
On my hap AC, on sticker `E01` is same as on port `eth0.2@eth0` and what is on `W01` is what is set in `wlan1`.

f00b4r0 replied:
OK so the VLANs have the correct MAC addresses but the root devices (eth0/eth1) do not. I don't know if that's ok. @adschm, what do you think?

rmounce replied:
`eth1` is SFP, this worked with 2 different GLC-T modules (which didn't work with RouterOS). I don't have any fibre transceivers to test with. I'll get it out this weekend & check the MAC address situation incl. SFP.

I used other devices (mpc85xx WDR4900) that have correct MAC address on VLAN ifaces but not on root dev so I think there is precedent there.

eosita replied:
Hi all. I've made a recap, hoping it could be useful:
My hAP ac with RouterOS immediatly after installation:

/interface ethernet> print
Flags: X - disabled, R - running, S - slave 
 #    NAME              MTU MAC-ADDRESS       ARP             SWITCH           
 0    ether1           1500 B8:69:F4:35:D6:74 enabled         switch1          
 1 RS ether2           1500 B8:69:F4:35:D6:75 enabled         switch1          
 2  S ether3           1500 B8:69:F4:35:D6:76 enabled         switch1          
 3  S ether4           1500 B8:69:F4:35:D6:77 enabled         switch1          
 4  S ether5           1500 B8:69:F4:35:D6:78 enabled         switch1          
 5 RS sfp1             1500 B8:69:F4:35:D6:79 enabled        

The sfp1 is equipped with a original S-31DLC20D (Single Mode, 20Km) and is connected to an identical module on another Mikrotik:
/interface ethernet> monitor 5
                      name: sfp1
                    status: link-ok
          auto-negotiation: done
                      rate: 1Gbps
               full-duplex: yes
           tx-flow-control: no
           rx-flow-control: no
               advertising: 
  link-partner-advertising: 
        sfp-module-present: yes
               sfp-rx-loss: no
                  sfp-type: SFP-or-SFP+
        sfp-connector-type: LC
       sfp-link-length-9um: 20000m
           sfp-vendor-name: MikroTik
    sfp-vendor-part-number: S-31DLC20D
       sfp-vendor-revision: 2.0
         sfp-vendor-serial: OL2110251169
    sfp-manufacturing-date: 21-10-25
            sfp-wavelength: 1310nm
           sfp-temperature: 40C
        sfp-supply-voltage: 3.197V
       sfp-tx-bias-current: 19mA
              sfp-tx-power: -6.278dBm
              sfp-rx-power: -7.649dBm
           eeprom-checksum: good
                    eeprom: 0000: 03 04 07 00 00 00 02 12  10 01 00 03 0d 00 14 c8  ........ ........
                            0010: 00 00 00 00 4d 69 6b 72  6f 54 69 6b 20 20 20 20  ....Mikr oTik    
                            0020: 20 20 20 20 00 00 00 00  53 2d 33 31 44 4c 43 32      .... S-31DLC2
                            0030: 30 44 20 20 20 20 20 20  32 2e 30 20 05 1e 00 39  0D       2.0 ...9
                            0040: 00 1a 00 00 4f 4c 32 31  31 30 32 35 31 31 36 39  ....OL21 10251169
                            0050: 20 20 20 20 32 31 31 30  32 35 20 20 68 b0 01 b5      2110 25  h...
                            0060: 00 00 08 0d 45 0d cc 22  34 fd 12 e8 02 fa 8c 6a  ....E.." 4......j
                            0070: 20 da 2c 00 00 00 00 00  00 00 00 00 97 cc 93 11   .,..... ........
                            0080: 64 00 d8 00 5f 00 dd 00  8c a0 6d 60 88 b8 71 48  d..._... ..m`..qH
                            0090: 9c 40 01 f4 88 b8 03 e8  18 a6 03 e8 13 94 04 eb  .@...... ........
                            00a0: 4d f1 00 14 27 10 00 28  00 00 00 00 00 00 00 00  M...'..( ........
                            00b0: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ........ ........
                            00c0: 00 00 00 00 3f 80 00 00  00 00 00 00 01 00 00 00  ....?... ........
                            00d0: 01 00 00 00 01 00 00 00  01 00 00 00 00 00 00 19  ........ ........
                            00e0: 29 16 7c 68 26 d6 09 10  08 fc ff ff ff ff 02 ff  ).|h&... ........
                            00f0: 00 40 ff ff 00 40 ff ff  ff ff ff ff ff ff ff 00  .@...@.. ........

/interface wireless> print      
Flags: X - disabled, R - running 
 0    name="wlan1" mtu=1500 l2mtu=1600 mac-address=B8:69:F4:35:D6:7B 
      arp=enabled interface-type=Atheros AR9300 mode=ap-bridge 
      ssid="MikroTik-35D67B" frequency=auto band=2ghz-b/g/n 
      channel-width=20/40mhz-Ce secondary-channel="" scan-list=default 
      wireless-protocol=802.11 vlan-mode=no-tag vlan-id=1 wds-mode=disabled 
      wds-default-bridge=none wds-ignore-ssid=no bridge-mode=enabled 
      default-authentication=yes default-forwarding=yes default-ap-tx-limit=0 
      default-client-tx-limit=0 hide-ssid=no security-profile=default 
      compression=no 

 1    name="wlan2" mtu=1500 l2mtu=1600 mac-address=B8:69:F4:35:D6:7A 
      arp=enabled interface-type=Atheros AR9888 mode=ap-bridge 
      ssid="MikroTik-35D67A" frequency=auto band=5ghz-a/n/ac 
      channel-width=20/40/80mhz-Ceee secondary-channel="" scan-list=default 
      wireless-protocol=802.11 vlan-mode=no-tag vlan-id=1 wds-mode=disabled 
      wds-default-bridge=none wds-ignore-ssid=no bridge-mode=enabled 
      default-authentication=yes default-forwarding=yes default-ap-tx-limit=0 
      default-client-tx-limit=0 hide-ssid=no security-profile=default 
      compression=no 

The two MAC Addresses reported on the sticker at the bottom of may hAP ac are the one that corresponds to interface "ether1" and interface "wlan1"
After taking these printouts I reinstalled the Daniel build and the output is exactly as before:

# ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether a2:01:f5:27:16:7c brd ff:ff:ff:ff:ff:ff
    inet6 fe80::a001:f5ff:fe27:167c/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether 52:c3:e4:c1:9a:9e brd ff:ff:ff:ff:ff:ff
5: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:35:d6:7a brd ff:ff:ff:ff:ff:ff
6: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether b8:69:f4:35:d6:7b brd ff:ff:ff:ff:ff:ff
7: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:35:d6:75 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fd02:f0d3:267b::1/60 scope global noprefixroute 
       valid_lft forever preferred_lft forever
    inet6 fe80::ba69:f4ff:fe35:d675/64 scope link 
       valid_lft forever preferred_lft forever
8: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether b8:69:f4:35:d6:75 brd ff:ff:ff:ff:ff:ff
9: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether b8:69:f4:35:d6:74 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::ba69:f4ff:fe35:d674/64 scope link 
       valid_lft forever preferred_lft forever

The MAC Addresses of eth0 and eth1 are unknown (to me) while eth0.1, eth0.2 and wlan0 correspond to the ones of RouterOS.
In Luci->Network->Switch only the 5 ethernet ports are visibile.

Looking around, I've discovered a similar issue in the official release of 21.02 for Mikrotik hEX S (device page: https://openwrt.org/toh/mikrotik/rb760igs):
"Fiber SFP is broken in 21.02. This patch in master fixes it, but it is not backported to 21.02, so you have to use snapshot builds if you use fiber SFP modules. Copper modules should work in 21.02."
I don't know if this is applicable also to hAP ac, but I would like to use the sfp port with fiber module, if possibile (the link is: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=54ed38d5ea)

rmounce commented on target/linux/ath79/dts/qca9558_mikrotik_routerboard-96x.dtsi:
```diff
@@ -0,0 +1,202 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "mikrotik,routerboard-962uigs-5hact2hnt", "qca,qca9558";
+	model = "MikroTik RouterBOARD 962UiGS-5HacT2HnT (hAP ac)";
```
Remove from `.dtsi`

danpawlik replied:
@rmounce thanks for replying. Could you rebase your change so the OpenWRT community will be able to merge?

rmounce replied:
Rebased and re-tested

danpawlik replied:
@rmounce thanks.
@f00b4r0  could you take a look on that PR please?

f00b4r0 reviewed (changes requested):
LGTM, except a few minor nits in the partition section of the DTS.
I notice some of these also appear in qca9533_mikrotik_routerboard-16m.dtsi and other mikrotik DTSes, you would have my blessings to submit a patch to clean them too. Otherwise I'll probably end up doing it :)

eosita commented:
Hi all. It's obvious I'm completely newbie on github, anyway I'll put my question: to test all the latest fixes do I need to dowloand a branch and try to compile it myself? Or there is somewhere a test build already available? If the answer is the former, where do I have to go to download the right sources? (I want to test particularly SFP fiber module, because all the others features works fine with last Daniel Pawlik test build). And, last question: when this release will be generally available? Thanks

rogerpueyo commented:
> Hi all. It's obvious I'm completely newbie on github, anyway I'll put my question: to test all the latest fixes do I need to dowloand a branch and try to compile it myself?

Yes, you need to fetch the repository _and_ the branch used for this particular pull request ([rmounce:ath79-rb962](https://github.com/rmounce/openwrt/tree/ath79-rb962)), and then compile it yourself. Unless somebody else is compiling it, which I am not aware of.

> And, last question: when this release will be generally available? Thanks

Once the pull request is merged, the _snapshot_ images for the device will start being compiled in one or two days.

danpawlik commented:
@eosita I will build new image tomorrow with newest code.

eosita commented:
@danpawlik I'll test this evening (my hAP ac is at home) and report results here. Thanks

eosita commented:
Sorry but I've tried a lot and I'm not been able to run the new image build by danpawlik.
I've tried with both reset button and reloading original routeros to change boot setting to "try-ethernet-once-then-nand", but the result is the same: the hAP ac remains "silent" on all five ethernet ports.
I'm sure that the image is loaded (I see data going through ethernet interface via "tcpdump") but nothing else happens.
To be sure, I've also re-loaded the previous image build by danpawlik and it works, so I think the the problem is in the new image itself.
Is there any other test I should try?

danpawlik commented:
@eosita strange. It works on my side, but I'm upgrading the system, not installing from scratch. Today I will create one more image. If this one will not work, could you check once again if the old image is working? ( image before @rmounce update the pull request). Maybe issue is somewhere else. Thanks!

eosita commented:
I follow your suggestions, Daniel. This is the results:
1. start from Routeros, boot with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
1.a install corresponding sysupgrade image: OK
1.b. sysupgrade to sysupgrade image snapshot r19115+1-41be1a2de2 of yesterday: OK
2. start from Routeros: boot with snapshot r19115+1-41be1a2de2  initramfs of yesterday: KO
3. boot wirh netboot (long press of reset) with snapshot r19115+1-41be1a2de2  initramfs of yesterday: KO
4. start from Routeros, boot with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
4.a install directly new sysupgrade image snapshot r19115+1-41be1a2de2 of yesterday: OK
5. start from Routeros: boot with snapshot r19128+1-46c5de5385 of today: KO
6. boot wirh netboot (long press of reset) with snapshot r19128+1-46c5de5385 of today: KO
7. boot wirh netboot (long press of reset) with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
7.a install directly new sysupgrade image snapshot r19128+1-46c5de5385 of today: OK
In short: the only iniramfs image working is the first one, but starting with it one can sysupgrado to any of the others. I've to check if fiber module works: it seems OK, but I have to check this better and now is very late: I'll do it tomorrow. Thanks

danpawlik commented:
> I follow your suggestions, Daniel. This is the results:
> 
> 1. start from Routeros, boot with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
>    1.a install corresponding sysupgrade image: OK
>    1.b. sysupgrade to sysupgrade image snapshot r19115+1-41be1a2de2 of yesterday: OK
> 2. start from Routeros: boot with snapshot r19115+1-41be1a2de2  initramfs of yesterday: KO
> 3. boot wirh netboot (long press of reset) with snapshot r19115+1-41be1a2de2  initramfs of yesterday: KO
> 4. start from Routeros, boot with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
>    4.a install directly new sysupgrade image snapshot r19115+1-41be1a2de2 of yesterday: OK
> 5. start from Routeros: boot with snapshot r19128+1-46c5de5385 of today: KO
> 6. boot wirh netboot (long press of reset) with snapshot r19128+1-46c5de5385 of today: KO
> 7. boot wirh netboot (long press of reset) with your "first" build 21.02.2 of 26/02/2022 initramfs image: OK
>    7.a install directly new sysupgrade image snapshot r19128+1-46c5de5385 of today: OK
>    In short: the only iniramfs image working is the first one, but starting with it one can sysupgrado to any of the others. I've to check if fiber module works: it seems OK, but I have to check this better and now is very late: I'll do it tomorrow. Thanks

That's strange.
I don't see so much differences in the commits proposed by @rmounce https://github.com/openwrt/openwrt/compare/7eb0b4bd08fdc3a0bebd13c0d7c92155547d2342..93ee7926ee65b07786e9037084b447e22cfd2240  (old commit vs new commit). 
I will do a new release with the old patch (7eb0b4bd08fdc3a0bebd13c0d7c92155547d2342), maybe issue is somewhere else.
It would be good if someone from OpenWRT community will check the diff.

f00b4r0 commented:
> I follow your suggestions, Daniel. This is the results:

What does "KO" exactly mean in your tests? Does the device reset or hang? is there some specific LED combination?
Having the output of serial console would help narrowing down what's going on.

joscha82 commented:
I have seen similar effects a few month ago while fiddling with the hardware. I believe the initramfs fails to boot correctly when exceeding a specific size on this hardware. I have not looked into it deeper, but i just saw that your new initramfs image is a bit bigger in size and you might have hit the sweet spot where it fails to boot.

eosita commented:
Hi all. What I mean with KO is that the device remains indefinitely (I wait up to 5 minutes) in a "undefined" state: Power (blue) led is ON and the eth1 (the connected ethernet) remains GREEN.
From the PC (linux) that serves the tftpboot I see (tcpdump) that the initramfs image has been transmitted then, on that interface, I only see ARP packets going thrugh. If, after some time (as I said, max 5 minutes), I connect the cable to another eth port, the only thing that happens is that the corresponding led becomes GREEN (eth1 led goes down) but I got no response on 192.168.1.1.
For ther serial console output, Mikrotik hAP ac has no "Device page" so I don't know is there is a way to access serial console and how to do it. Looking to the router there is no console port, the only non ethernet port is the usb.
Honestly, like @joscha82, I've noticed the different dimension of initramfs images (the running one is little more that 5Mb, the others are 6.5Mb) and asked myself how can the @mounce commits change so much, anyway I've no idea if this could be the root cause on non-booting initramfs

f00b4r0 commented:
OK, it's not @rmounce commits that cause this change: you were testing a 21.02 against a *snapshot* build. The latter has newer kernel and is generally bigger, so I suspect that yes, we are hitting a size issue here as has been seen on other devices too.

danpawlik commented:
@eosita I create one more release that contains builds without any additional package (base on OpenWRT snapshot config).
It does not include even luci.
One image is with patch before rebase by @rmounce and second after. If you have time, please check them (I'm not able to do it for few days).

Thanks

eosita commented:
Sorry, but no good news. The only initramfs image that boots is the one Daniel build from 21.02 on Feb, 12 (in other words, none of the snaphost images boot). On the other side, all of the sysupgrade images work (also the last two build by Daniel on friday), obviously loaded from the unique initramfs image that boots (by the way, i do not see any difference between last two from the usage point of view).
I've also tested the fiber sfp module and I've been able to light it up as "eth1" with all sysupgrade images (I've still some problem to manage it in bridging and/or routing, but I think it's may poor knowledge, because if I assign an IP to eth1, the device at the other side answer)

f00b4r0 commented:
@danpawlik @eosita the problem likely isn't the size of the image, it's the size of the kernel.

danpawlik commented:
> @danpawlik @eosita the problem likely isn't the size of the image, it's the size of the kernel.

yes, I just want to confirm that nothing others can be related to it. I'm new in OpenWRT and by doing one more test it is clear now. Kernel 5.10 is only one right now available for future releases, right? Is there any patch that is making kernel size smaller and it require test/review?

eosita commented:
Hi all. Just to understand: is there a next step waiting to be done?
Seen that booting with the 21.02 initramfs image permits to sysupgrade to any version (and also that sysupgrade is availabale from any installed version towards at least all the images I tested), could we get something like a "stable" image (or snapshot image), but that coudl be installed and for which one can add modules and other packages?

danpawlik commented:
could you update your commit @rmounce  please?

rmounce commented:
Fixed DTS partition nits.

hauke commented:
Thank you for your patch, I applied it to master in c2140e32ce32b9cc60f7d408e20bdf45dce6a634.

eosita commented:
Excuse me, could someone put me in the right direction of learning how to build a release from the updated master? A release for which, after installation, modules could be added? Thanks in advance

turtureanu commented:
> Excuse me, could someone put me in the right direction of learning how to build a release from the updated master? A release for which, after installation, modules could be added? Thanks in advance

You can try following: https://openwrt.org/docs/guide-developer/toolchain/use-buildsystem

Make sure to include the following packages if they aren't already present in the device build profile `kmod-ath kmod-ath9k kmod-ath9k-common ath10k-firmware-qca9887-ct ath10k-firmware-qca988x-ct`, these make sure the 5GHz chip works. If I'm not mistaken, you can add then from the `make menuconfig`.

If you're interested in using the generated snapshot release please read this:  https://openwrt.org/releases/snapshot

f00b4r0 commented:
> hi, usb is not working for me, tested on two devices. dmesg shows nothing when i plug in a stick oder usb-modem.
> tested on snapshot Tue Jun 28 03:02:27 2022

you need to turn on USB power gpio.

evil-god commented:
> > hi, usb is not working for me, tested on two devices. dmesg shows nothing when i plug in a stick oder usb-modem.
> > tested on snapshot Tue Jun 28 03:02:27 2022
> 
> you need to turn on USB power gpio.

thx for the hint XD, i will try it later this day

danpawlik commented:
> > > hi, usb is not working for me, tested on two devices. dmesg shows nothing when i plug in a stick oder usb-modem.
> > > tested on snapshot Tue Jun 28 03:02:27 2022
> > 
> > 
> > you need to turn on USB power gpio.
> 
> thx for the hint XD, i will try it later this day

Please leave comment with a command if you have done it.

evil-god commented:
dongle is now shown in dmesg

`[  209.961081] usb 1-1: new high-speed USB device number 2 using ehci-platform
[  210.687694] usb 1-1: USB disconnect, device number 2
[  211.461139] usb 1-1: new high-speed USB device number 3 using ehci-platform
`

keiser1080 commented:
@danpawlik is there a way to test  your 22.03 build without flashing?
i use the 19..07.10 but the poe on port 5 do not work.
Hi have triyed all your initramfs +  [22.03.0-rc5](https://downloads.openwrt.org/releases/22.03.0-rc5/targets/ath79/mikrotik/openwrt-22.03.0-rc5-ath79-mikrotik-mikrotik_routerboard-962uigs-5hact2hnt-initramfs-kernel.bin)

@eosita how to you test the poe on port 5 ?
with the 19..07.10  i plug the a poe eth cable from a 802.3.af swith and i connect a 802.3.af access point but that don't work.
I can change the color of the led to red by changing the gpio 3.

rmounce commented:
@keiser1080 the hAP ac only supports passive PoE not 802.3af.

rmounce commented:
@keiser1080 I don’t think the hAP ac has the hardware to negotiate 802.3af either as a PD on port 1 or PSE on port 5. You will need a separate injector if your downstream device only supports 802.3af.

This is not an OpenWrt limitation, just MikroTik cost cutting.

danpawlik commented:
> 

@keiser1080 hey, I stopped creating own builds for that device. 
You can try initramfs image as you do on moving from RouterOS to OpenWRT.
